File: app/src/main/java/com/damhoe/fieldlines/ui/EditChargesFragment.java
package com.damhoe.fieldlines.ui;

import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.PopupMenu;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.view.ActionMode;
import androidx.appcompat.view.menu.MenuBuilder;
import androidx.core.view.MenuProvider;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.damhoe.fieldlines.domain.Charge;
import com.damhoe.fieldlines.domain.ChargeList;
import com.damhoe.fieldlines.ui.EditableChargeAdapter.ChargeViewHolder;
import com.example.fieldlines.R;
import com.example.fieldlines.databinding.FragmentEditChargesBinding;
import com.google.android.material.snackbar.Snackbar;

import java.util.Objects;


public class EditChargesFragment extends Fragment implements NotifyItemClickListener {

    FragmentEditChargesBinding binding;
    SharedViewModel viewModel;
    ActionMode actionMode;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_edit_charges, container, false);
        initializeMenu();
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        viewModel = new ViewModelProvider(requireActivity()).get(SharedViewModel.class);

        viewModel.getCharges().observe(getViewLifecycleOwner(), this::updateAdapter);

        // Setup adapter
        EditableChargeAdapter adapter = new EditableChargeAdapter(this);
        adapter.updateCharges(viewModel.getCharges().getValue());
        binding.chargesRecyclerView.setAdapter(adapter);
        binding.chargesRecyclerView.setLayoutManager(new LinearLayoutManager(requireContext()));
        binding.chargesRecyclerView.addItemDecoration(new EditableChargeAdapter.ItemDecoration(4));
    }

    private void updateAdapter(ChargeList charges) {
        EditableChargeAdapter adapter = (EditableChargeAdapter)binding.chargesRecyclerView.getAdapter();
        if (adapter != null) {
            adapter.updateCharges(charges);
        }
    }

    private void initializeMenu() {
        requireActivity().addMenuProvider(new MenuProvider() {
            @Override
            public void onCreateMenu(@NonNull Menu menu, @NonNull MenuInflater menuInflater) {
                menu.clear();
                menuInflater.inflate(R.menu.menu, menu);
                menu.removeItem(R.id.menu_library);
                menu.removeItem(R.id.menu_about);
                menu.findItem(R.id.menu_help).setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM);
            }

            @Override
            public boolean onMenuItemSelected(@NonNull MenuItem menuItem) {
                if (menuItem.getItemId() == R.id.menu_help) {
                    findNavController().navigate(R.id.action_editChargesFragment_to_helpFragment);
                }
                return true;
            }
        }, getViewLifecycleOwner());
    }

    private NavController findNavController() {
        return Navigation.findNavController(requireActivity(), R.id.nav_host_fragment);
    }

    @Override
    public void notifyChargeClicked(int position, Charge charge) {
        ChargeViewHolder viewHolder = (ChargeViewHolder)
                binding.chargesRecyclerView.findViewHolderForAdapterPosition(position);

        if (viewHolder == null) {
            return;
        }

        PopupMenu popup = new PopupMenu(requireContext(), viewHolder.buttonMore);
        popup.getMenuInflater().inflate(R.menu.menu_charge, popup.getMenu());
        popup.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem menuItem) {
                if (menuItem.getItemId() == R.id.edit) {
                    showEditChargeDialog(position, charge);
                    return true;
                }
                if (menuItem.getItemId() == R.id.delete) {
                    Charge removedCharge = viewModel.removeCharge(position);
                    Snackbar.make(binding.getRoot(), "Removed charge", Snackbar.LENGTH_SHORT)
                            .setAction("Undo", view -> {
                                viewModel.addCharge(position, removedCharge);
                            })
                            .show();
                    return true;
                }
                return false;
            }
        });
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            popup.setForceShowIcon(true);
        }
        popup.show();
    }

    private void showEditChargeDialog(int position, Charge charge) {
        new ChargeAlertDialogFactory(requireContext(), new ChargeRequestListener() {
            @Override
            public void notifyChargeRequest(int x, int y, double amount) {
                viewModel.updateCharge(position, x, y, amount);
            }
        }).createEditChargeDialog(charge).show();
    }
}

File: app/src/main/java/com/damhoe/fieldlines/charges/presentation/EditChargesFragment.kt
package com.damhoe.fieldlines.charges.presentation

import android.graphics.PointF
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.PopupMenu
import androidx.core.view.MenuProvider
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.NavController
import androidx.navigation.Navigation.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import com.damhoe.fieldlines.charges.application.PointChargeDialogBuilder
import com.damhoe.fieldlines.charges.domain.PointCharge
import com.damhoe.fieldlines.charges.presentation.ChargeItemAdapter.ChargeViewHolder
import com.damhoe.fieldlines.field.domain.Field
import com.damhoe.fieldlines.field.presentation.FieldViewModel
import com.example.fieldlines.R
import com.example.fieldlines.databinding.FragmentEditChargesBinding
import com.google.android.material.divider.MaterialDividerItemDecoration
import com.google.android.material.snackbar.Snackbar

class EditChargesFragment : Fragment() {
    private lateinit var binding: FragmentEditChargesBinding
    private val viewModel: FieldViewModel by viewModels ({ requireActivity() })

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        // Inflate the layout for this fragment
        binding = DataBindingUtil.inflate(inflater,
            R.layout.fragment_edit_charges, container, false)

        addMenu()

        setupAddChargeButton()

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel.field.observe(viewLifecycleOwner) { field: Field -> updateAdapter(field) }

        // Setup adapter
        val adapter = ChargeItemAdapter {
            position, charge -> handleChargeSelection(position, charge)
        }

        viewModel.field.value?.let { adapter.updateCharges(it.pointCharges) }

        binding.chargesRecyclerView.run {
            this.adapter = adapter
            layoutManager = LinearLayoutManager(requireContext())
            val dividerItemDecoration = MaterialDividerItemDecoration(context,
                MaterialDividerItemDecoration.VERTICAL)
            addItemDecoration(dividerItemDecoration)
        }
    }

    private fun updateAdapter(field: Field) {
        (binding.chargesRecyclerView.adapter as ChargeItemAdapter)
            .updateCharges(field.pointCharges)
    }

    private fun addMenu() {
        requireActivity().addMenuProvider(object : MenuProvider {
            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                menu.run {
                    clear()
                }
            }

            override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
               return false
            }
        }, viewLifecycleOwner)
    }

    private fun findNavController(): NavController {
        return findNavController(requireActivity(), R.id.nav_host_fragment)
    }

    private fun handleChargeSelection(position: Int, charge: PointCharge) {
        val viewHolder = binding.chargesRecyclerView
            .findViewHolderForAdapterPosition(position) as ChargeViewHolder
        val popup = PopupMenu(requireContext(), viewHolder.binding.buttonMore)
        popup.menuInflater.inflate(R.menu.menu_charge, popup.menu)
        popup.setOnMenuItemClickListener(PopupMenu.OnMenuItemClickListener { menuItem ->
            if (menuItem.itemId == R.id.edit) {
                showEditChargeDialog(position, charge)
                return@OnMenuItemClickListener true
            }
            if (menuItem.itemId == R.id.delete) {
                val result = viewModel.removePointChargeAt(position)
                result.getOrNull()?.let { pointCharge ->
                    Snackbar.make(binding.root, "Removed charge", Snackbar.LENGTH_SHORT)
                        .setAction("Undo") {
                            viewModel.addPointChargeAt(position, pointCharge)
                        }
                        .show()
                }

                return@OnMenuItemClickListener true
            }
            false
        })
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            popup.setForceShowIcon(true)
        }
        popup.show()
    }

    private fun setupAddChargeButton() {
        binding.buttonAdd.setOnClickListener {

            val defaultCoordinates = PointF(0f, 0f)
            val defaultCharge = 1.0

            PointChargeDialogBuilder(requireContext())
                .setTitle("Create charge")
                .setCharge(defaultCharge)
                .setPoint(defaultCoordinates)
                .setPositiveButton("Create") { x, y, charge ->
                    viewModel.createPointCharge(x, y, charge).onFailure {
                        Snackbar.make(
                            binding.root,
                            "Failed: ${it.message}",
                            Snackbar.LENGTH_SHORT
                        )
                            .setAction("OK") {  }
                            .show()
                    }
                }
                .setNegativeButton("Cancel") { d, _ -> d.cancel() }
                .create().show()
        }
    }

    private fun showEditChargeDialog(position: Int, pointCharge: PointCharge) {
        PointChargeDialogBuilder(requireContext())
            .setTitle("Edit charge")
            .setPoint(pointCharge.position)
            .setCharge(pointCharge.charge)
            .setPositiveButton("Save") { x, y, charge ->
                viewModel.updateCharge(position, x, y, charge).onFailure {
                    Snackbar.make(
                        binding.root,
                        "Failed: ${it.message}",
                        Snackbar.LENGTH_SHORT
                    )
                        .setAction("OK") {  }
                        .show()
                }
            }
            .setNegativeButton("Cancel") { d, _ -> d.cancel() }
            .create().show()
    }
}
